src/Gui/propertyeditor/PropertyItem.cpp

```diff
@@ -2882,18 +2905,91 @@ void PropertyEnumItem::setEditorData(QWidget *editor, const QVariant& data) cons
         }
     }
 
-    QComboBox *cb = qobject_cast<QComboBox*>(editor);
-    if (!commonModes.isEmpty()) {
-        cb->clear();
+    if (commonModes.isEmpty())
+        return nullptr;
+
+    int index = -1;
+    std::shared_ptr<EnumItems> enumItems;
+    for (auto &mode : commonModes) {
+        ++index;
+        auto fields = mode.split(QStringLiteral("|"));
+        if (!enumItems && fields.size() <= 1)
+            continue;
+        if (!enumItems) {
+            enumItems = std::make_shared<EnumItems>();
+            for (int i=0; i<index; ++i)
+                enumItems->emplace_back(commonModes[i], mode);
+        }
+        auto children = enumItems;
+        int j = -1;
+        for (auto &field : fields) {
+            ++j;
+            field = field.trimmed();
+            auto it = children->end();
+            if (field.isEmpty()) {
+                if (children->size())
+                    --it;
+                else
+                    continue;
+            } else {
+                it = std::find_if(children->begin(), children->end(),
+                    [&field](const EnumItem &item) {
+                        return item.text == field;
+                    });
+                if (it == children->end())
+                    it = children->emplace(children->end(), field, mode);
+            }
+            if (j + 1 == (int)fields.size())
+                break;
+            if (!it->children)
+                it->children = std::make_shared<EnumItems>();
+            children = it->children;
+        }
+    }
+
+    if (!enumItems) {
+        QComboBox *cb = new QComboBox(parent);
+        cb->setFrame(false);
+        cb->setDisabled(isReadOnly());
         cb->addItems(commonModes);
+        QObject::connect(cb, SIGNAL(activated(int)), receiver, method);
+        return cb;
+    }
+
+    auto button = new PropertyEnumButton(parent);
+    button->setDisabled(isReadOnly());
+    QMenu *menu = new QMenu(button);
+    for (auto &item : *enumItems)
+        item.populate(menu);
+    button->setMenu(menu);
+    QObject::connect(menu, &QMenu::aboutToShow, this, [=]() {
+        menu->setMinimumWidth(button->width());
+    });
+    QObject::connect(menu, &QMenu::triggered, this, [=](QAction *action) {
+        button->setText(action->data().toString());
+        button->picked();
+    });
+    QObject::connect(button, SIGNAL(picked()), receiver, method);

```

Comment:
Prefer new-style connection so we get compile-time checks.

Reply:
Same as above